### PR TITLE
Bugs in python fit functions

### LINKED ISSUE #22365: Bugs in python fit functions
1. Newly defined python fit function cannot be used with the new function wrappers, as demonstrated in script below:
```
class Example1DFunction(IFunction1D):

    def init(self):
        self.declareParameter("C", 0.0)

    def function1D(self, xvals):
        c = self.getParameterValue("C")
        return c*xvals

FunctionFactory.subscribe(Example1DFunction)

my_func = Gaussian()
# Correct: prints out that function is class 'mantid.fitfunctions.FunctionWrapper'
print('Type of my_func: {0}'.format(type(my_func)))

my_func = Example1DFunction()
# Wrong: Does not print out that funciton is class 'mantid.fitfunctions.FunctionWrapper'
print('Type of my_func: {0}'.format(type(my_func)))

# and because of the above this fails
my_func = Example1DFunction(C=1.0)
print(my_func(1.0))
```
2. `IFunction` doesn't export access to the fit errors.
3. It looks as if the fit functions defined in `PythonInterface/plugins/functions/` don't get wrapped in FunctionWrapper.
```
print(Guinier())
```
4. `MultiDomanFunction()` must set `NumDeriv=True` either by default or automatically when needed.

### PR DESCRIPTION
**Pull request by mantid-roman**
**Description of work**

1. Overriden `IFunction1D`'s constructor such that it returns a `FunctionWrapper` when called from a script. FunctionFactory still creates normal fit function instance.
2. Exposed `IFunction.getError` method to python.
3. Wrap functions defined in the plugins directory.
4. Set `NumDeriv` of `MultiDomainFunction` to `true` by default.

**To test:**

1. Run the script form the issue description.
2. Try using `IFunction.getError` method.
3. Try creating functions defined in the plugins directory.
4. Run this script:
```
x = np.linspace(-10, 10)
y1 = np.exp(-(x-2)**2)
y2 = 2*np.exp(-x**2)
y3 = 3*np.exp(-(x+2)**2)

ws1 = CreateWorkspace(x, y1)
ws2 = CreateWorkspace(x, y2)
ws3 = CreateWorkspace(x, y3)

mdf = MultiDomainFunction(Gaussian(), Gaussian(), Gaussian(), Global=["Sigma"])
res = Fit(mdf, InputWorkspace=ws1, InputWorkspace_1=ws2, InputWorkspace_2=ws3)
f = res.Function

print(f[0].Sigma)  # 0.707106780886
print(f[1].Sigma)  # 0.707106780886
print(f[2].Sigma)  # 0.707106780886
print(f[0].Height)  # 1
print(f[1].Height)  # 2
print(f[2].Height)  # 3
print(f[0].PeakCentre)  # 2
print(f[1].PeakCentre)  # 0
print(f[2].PeakCentre)  # -2
```

Fixes #22365

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Anders-Markvardsen** [comment]
@mantid-roman please note the merge conflicts

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 1]
```diff
@@ -397,6 +409,8 @@ def initByName(self, name, *args, **kwargs):
                        composite function itself.
         """
         if len(args) == 1 and  not isinstance(args[0], FunctionWrapper):
+            if isinstance(args[0], str):
+                raise RuntimeError('Whaaat? {}'.format(args[0]))
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: Can we have a better error message, please?

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 2]
```diff
@@ -731,25 +755,32 @@ def _create_wrapper_function(name):
...
+_OldIFunction1D = IFunction1D
+
+
+class _NewIFunction1D(_OldIFunction1D):
+    """Overriden IFunction1D class that allows creation of
+    FunctionWrappers for newly defined fit functions.
+    """
+
+    def __new__(cls, _is_called_by_function_factory=False, **kwargs):
+        if _is_called_by_function_factory or (hasattr(cls.__init__, '__code__') and
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: I think we could do with some comments here to explain what is going on and how the boolean connects back to the factory as the second part of the `if` is not completely clear here.

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 3]
```diff
@@ -731,25 +755,32 @@ def _create_wrapper_function(name):
...
 
 
-def _create_wrappers_for_all_fit_functions():
+def _attach_wrappers(source_module):
     for name in FunctionFactory.getFunctionNames():
-        _create_wrapper_function(name)
+        if name not in _do_not_wrap:
+            setattr(source_module, name, _create_wrapper_function(name))
+
+
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: I think some high-level comments about the dance that is happening here would be useful - essentially explaining why we are doing this.

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 4]
```diff
@@ -731,25 +755,32 @@ def _create_wrapper_function(name):
...
+_OldIFunction1D = IFunction1D
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: How about `_ExportedIFunction1D` as a name?

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 5]
```diff
@@ -731,25 +755,32 @@ def _create_wrapper_function(name):
...
+
+
+class _NewIFunction1D(_OldIFunction1D):
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: How about `_IFunction1DWrapperCreator`?

File: Framework/PythonInterface/plugins/functions/PrimStretchedExpFT.py [hunk 6]
```diff
@@ -35,8 +35,8 @@
 
 class PrimStretchedExpFT(IFunction1D):
     # pylint: disable=super-on-old-class
-    def __init__(self):
-        super(self.__class__, self).__init__()
+    def __init__(self, _is_called_by_function_factory):
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/functions/PrimStretchedExpFT.py]: Is there anyway to shield users from this? It's going to be a tough one to explain what that argument means.

Would a fallback option be to use something like
```python
def __init__(self, *args):
    super(self.__class__, self).__init__(*args)
```

to be less explicit in the future if we need to provide flexibility for other arguments?

File: Framework/PythonInterface/mantid/api/src/Exports/FunctionFactory.cpp [hunk 7]
```diff
@@ -24,6 +24,39 @@ using namespace boost::python;
 
 GET_POINTER_SPECIALIZATION(FunctionFactoryImpl)
 
+namespace Mantid {
+namespace PythonInterface {
+
+/// Specialization for IFunction. It needs to pass a bool to the created
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/FunctionFactory.cpp]: This is good - I like that we can hide this from users. Could you update the comment to explain a little bit about how the mechanism works?

File: Framework/PythonInterface/mantid/api/src/FitFunctions/IPeakFunctionAdapter.cpp [hunk 8]
```diff
@@ -12,8 +12,10 @@ using namespace boost::python;
 /**
  * Construct the "wrapper" and stores the reference to the PyObject
  * @param self A reference to the calling Python object
+ * @param isCreatedByFactory FunctionFactory should set this agument to true
+ *  when creatting the function.
  */
-IPeakFunctionAdapter::IPeakFunctionAdapter(PyObject *self)
+IPeakFunctionAdapter::IPeakFunctionAdapter(PyObject *self, bool)
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/FitFunctions/IPeakFunctionAdapter.cpp]: Is the `bool` still required here?

File: Framework/PythonInterface/mantid/api/src/Exports/IPeakFunction.cpp [hunk 9]
```diff
@@ -10,7 +10,8 @@ using namespace boost::python;
 void export_IPeakFunction() {
   class_<IPeakFunction, bases<IFunction1D>,
          boost::shared_ptr<IPeakFunctionAdapter>, boost::noncopyable>(
-      "IPeakFunction")
+      "IPeakFunction", "Base class for peak Fit functions", init<>())
+      .def(init<bool>())
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/IPeakFunction.cpp]: Is the bool still required here?

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 10]
```diff
@@ -772,6 +775,15 @@ def _attach_wrappers(source_module):
 class _IFunction1DWrapperCreator(_ExportedIFunction1D):
     """Overriden IFunction1D class that allows creation of
     FunctionWrappers for newly defined fit functions.
+
+    User defined function inherit from this class.
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/fitfunctions.py]: :+1: Thanks

**Comment by Anders-Markvardsen** [comment]
When I run

```
import numpy as np

my_func = Lorentzian(Amplitude=1.0, FWHM=1.0, PeakCentre = 0.0)
print(my_func)
print('Type of my_func: {0}'.format(type(my_func)))

my_func(1.0)
my_array = np.array([[0, 1],[2, 3]])
my_func(my_array)

class MyScaleFunction(IFunction1D):

    def init(self):
        self.declareParameter("C", 2.0)

    def function1D(self, xvals):
        # Access current values during the fit
        c = self.getParameterValue("C")

        return c*numpy.sqrt(xvals)

FunctionFactory.subscribe(MyScaleFunction)

my_func = MyScaleFunction()
print(my_func)
print('Type of my_func: {0}'.format(type(my_func)))

my_func(1.0)
my_array = np.array([[0, 1],[2, 3]])
my_func(my_array)
```

I get the messages and the error:

```
name=Lorentzian,Amplitude=1,PeakCentre=0,FWHM=1
Type of my_func: <class 'mantid.fitfunctions.FunctionWrapper'>
name=MyScaleFunction,C=2
Type of my_func: <class 'mantid.fitfunctions.FunctionWrapper'>
RuntimeError: global name 'numpy' is not defined
  at line 20 in 'New script'
  at line 28 in 'New script'
```

Hence the custom function appears to get registere, but appears to behave differently to a build in function when you pass it arguments

**Comment by mantid-roman** [comment]
@Anders-Markvardsen Sorry, I cannot see the difference.

**Comment by Anders-Markvardsen** [comment]
Do you mean you don't see: 

```
RuntimeError: global name 'numpy' is not defined
  at line 20 in 'New script'
  at line 28 in 'New script'
```
?

**Comment by mantid-roman** [comment]
This is because numpy is imported as np. Use np instead of numpy.

**Comment by Anders-Markvardsen** [comment]
What a relief, now work with this change. Recommendation - don't do testing late in the afternoon! I would like one final look and will do this before I leave today

File: Framework/Algorithms/src/FitPeak.cpp [hunk 11]
```diff
@@ -744,6 +744,8 @@ double FitOneSinglePeak::fitFunctionMD(IFunction_sptr fitfunc,
   boost::shared_ptr<MultiDomainFunction> funcmd =
       boost::make_shared<MultiDomainFunction>();
 
+  funcmd->setAttributeValue("NumDeriv", false);
```
Comment (Anders-Markvardsen) [inline Framework/Algorithms/src/FitPeak.cpp]: Please explain why you are setting NumDeriv to false, perhaps also an explanation why you changed the default in the constructor in MultiDomainFunction.cpp to true. thanks

**Comment by Anders-Markvardsen** [comment]
What about release notes - good to add for this work I think

**Review by Anders-Markvardsen — APPROVED** [review]
Looks good. Can't fail it and unit test coverage etc. thanks

**Comment by mantid-roman** [comment]
@Anders-Markvardsen @martyngigg Do we want to have this decorator?
```
from mantid.simpleapi import (FunctionFactory, IFunction1D)


def make_ff(*args, **kwargs):

    defaults = kwargs

    def maker(fun):

        class WrapperFFClass(IFunction1D):
            param_names = fun.__code__.co_varnames[1:fun.__code__.co_argcount]

            def init(self):
                for name in self.param_names:
                    self.declareParameter(name, defaults.get(name, 0.0))

            def function1D(self, xvals):
                params = [self.getParameterValue(name) for name in self.param_names]
                return fun(xvals, *params)

        cls = WrapperFFClass
        cls.__name__ = fun.__name__
        FunctionFactory.subscribe(cls)
        return cls

    if len(args) == 1 and hasattr(args[0], '__call__'):
        return maker(args[0])
    return maker


@make_ff
def my_fun(x, a, b):
    return x*a + b


print(my_fun)
f = my_fun(a=2, b=1)
print(f)
print(f([2, 4, 5]))


@make_ff(a=1)
def other_fun(x, a, b):
    return x*a + b


o = other_fun()
o.b = 2
print(o)
print(o([2, 4, 5]))

g = f + o
print(g)
print(g([2, 4, 5]))
```

**Comment by Anders-Markvardsen** [comment]
ooohhh yes please, but I would call it something different form make_ff, like for example make_fit_function. I think this would a nice additional way of defining fit functions

**Comment by mantid-roman** [comment]
OK. I'll open a new issue for it.

**Comment by Anders-Markvardsen** [comment]
thanks and don't hesitate to assign me as tester

**Comment by martyngigg** [comment]
@mantid-roman That sounds good. Does it work with `Fit` like this?

Can you put quite a few details into the intent into the new issue so we can get some other opinions on it?

**Comment by mantid-roman** [comment]
It does work with Fit. It also supposed to work with `scipy.optimize.curve_fit` but for some reason it doesn't with python-2 (for me at least) but I think it should work with python-3.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantid-roman I think we'd have to explore why it doesn't work with Python 2 first - we're stuck with that for a while :(

**Comment by martyngigg** [comment]
Build failures here are unrelated.

**Comment by mantid-roman** [comment]
@martyngigg `curve_fit` in python 2 wants a normal python function:
```
TypeError: <__main__.A object at 0x00000000126116A0> is not a Python function
```

Python 3 is happy with any callable object.

**Pull request merged**